canon/src/implementations.rs

```diff
@@ -216,36 +175,105 @@ macro_rules! canon_tuple_impls {
 
             fn encoded_len(&self) -> usize {
-                let mut n = 0;
+                0 $( + self.$idx.encoded_len())*
+            }
+        }
+    };
+}
 
-                $(n += self.$idx.encoded_len();)+
-                n
+tuple! { A B, 0 1 }
+tuple! { A B C, 0 1 2 }
+tuple! { A B C D, 0 1 2 3 }
+tuple! { A B C D E, 0 1 2 3 4 }
+tuple! { A B C D E F, 0 1 2 3 4 5 }
+tuple! { A B C D E F G, 0 1 2 3 4 5 6 }
+tuple! { A B C D E F G H, 0 1 2 3 4 5 6 7 }
+tuple! { A B C D E F G H I, 0 1 2 3 4 5 6 7 8 }
+tuple! { A B C D E F G H I J, 0 1 2 3 4 5 6 7 8 9 }
+tuple! { A B C D E F G H I J K, 0 1 2 3 4 5 6 7 8 9 10 }
+tuple! { A B C D E F G H I J K L, 0 1 2 3 4 5 6 7 8 9 10 11 }
+tuple! { A B C D E F G H I J K L M, 0 1 2 3 4 5 6 7 8 9 10 11 12 }
+tuple! { A B C D E F G H I J K L M N, 0 1 2 3 4 5 6 7 8 9 10 11 12 13 }
+tuple! { A B C D E F G H I J K L M N O, 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 }
+tuple! { A B C D E F G H I J K L M N O P, 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 }
+
+macro_rules! array {
+    ($n:expr) => {
+        impl<T, S> Canon<S> for [T; $n]
+        where
+            T: Canon<S> + Sized,
+            S: Store,
+        {
+            fn write(&self, sink: &mut impl Sink<S>) -> Result<(), S::Error> {
+                for i in 0..$n {
+                    self[i].write(sink)?;
+                }
+                Ok(())
+            }
+
+            fn read(source: &mut impl Source<S>) -> Result<Self, S::Error> {
+                let mut array = arrayvec::ArrayVec::new();
+
+                for _ in 0..$n {
+                    array.push(T::read(source)?);
+                }
+
+                Ok(array
+                    .into_inner()
+                    .map_err(|_| ())
+                    .expect("Array not full after N pushes"))
+            }
+
+            fn encoded_len(&self) -> usize {
+                let mut len = 0;
+                for i in 0..$n {
+                    len += self[i].encoded_len();
+                }
+                len
             }
         }
     };
 }
 
-canon_tuple_impls! { A B, 0 1 }
-canon_tuple_impls! { A B C, 0 1 2 }
-canon_tuple_impls! { A B C D, 0 1 2 3 }
-canon_tuple_impls! { A B C D E, 0 1 2 3 4 }
-canon_tuple_impls! { A B C D E F, 0 1 2 3 4 5 }
-canon_tuple_impls! { A B C D E F G, 0 1 2 3 4 5 6 }
-canon_tuple_impls! { A B C D E F G H, 0 1 2 3 4 5 6 7 }
-canon_tuple_impls! { A B C D E F G H I, 0 1 2 3 4 5 6 7 8 }
-canon_tuple_impls! { A B C D E F G H I J, 0 1 2 3 4 5 6 7 8 9 }
-canon_tuple_impls! { A B C D E F G H I J K, 0 1 2 3 4 5 6 7 8 9 10 }
-canon_tuple_impls! { A B C D E F G H I J K L, 0 1 2 3 4 5 6 7 8 9 10 11 }
-canon_tuple_impls! { A B C D E F G H I J K L M, 0 1 2 3 4 5 6 7 8 9 10 11 12 }
-canon_tuple_impls! { A B C D E F G H I J K L M N, 0 1 2 3 4 5 6 7 8 9 10 11 12 13 }
-canon_tuple_impls! { A B C D E F G H I J K L M N O, 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 }
-canon_tuple_impls! { A B C D E F G H I J K L M N O P, 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 }
+array!(0);
```

Comment:
Even though the API is nicer, i was worried about not understanding the unsafe 100%, and since we had segfaults, this is an ok compromise for now. We should change it back (but to conservative, as you recommended) later.

Reply:
For worries about unsafe that i don't 100% understand. We should add it back in later (in the conservative version as you suggsted)